### PR TITLE
Use pydata-sphinx-theme

### LINKED ISSUE #11520: DOC: pydata sphinx theme
### Description
I would propose to use the PyData Sphinx Theme for the documentation. This theme is used by most Python tools from NumFOCUS. Examples include: NumPy, Pandas, SciPy (devdocs) and also Matplotlib (WIP).

Let me know if this is something which could interest you. I did the work for [SciPy](https://github.com/scipy/scipy/pull/13724) and can do a PR here.

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to supersede #14477 and fix #11520 . I moved some common stuff upstream to `sphinx-astropy` and managed to get this working without moving all the docs around.

Rendered doc: https://astropy--14867.org.readthedocs.build/en/14867/

Pros:

* Dark mode!
* Use a theme that is also used by other Python libraries (e.g., numpy, matplotlib). It has mobile support, is quite responsive, and did I say dark mode?

Cons (I don't think it is a deal breaker):

* https://github.com/pydata/pydata-sphinx-theme/issues/1072
* https://github.com/pydata/pydata-sphinx-theme/issues/1269 (though this problem also exists in the old theme, just more hidden under the tiny bottom right menu)
* https://github.com/astropy/sphinx-automodapi/issues/167
* https://github.com/matplotlib/matplotlib/issues/25983
* It takes getting used to (though pretty intuitive once you realize left/right sidebars show different things).
* Have to create a several new pages to make the sidebars work for us (but better than moving files around).

Requires:

* https://github.com/astropy/sphinx-astropy/pull/59

TODO:

- [x] Wait for sphinx-astropy release with required feature.
- [x] Remove hack in `setup.cfg`.

Follow-up action items after merge:

- Send out communication about theme change and map the old user guide access to new (maybe screenshots of both and some annotation between them?).

### CODE REVIEW

**Comment by pllim** [comment]
pre-commit.ci autofix

**Comment by pllim** [comment]
RTD is taking way too long and it might time out. Maybe it doesn't like me specifying my own branch of sphinx-astropy. Not sure what to do here to make it happy. 🐔 🥚

**Comment by pllim** [comment]
Thanks for all your help, @drammock ! If you want, I can add you as co-author on the second commit. Just have to give me the info I can link to your GitHub account (`Name <email>`).

**Comment by saimn** [comment]
Thanks for working on that @pllim :)
There is a thing that is missing compared to the old PR (maybe due to the toctree tweaks): the theme can show the toctree of the subpackage when browsing a subpage, in the left bar, e.g.
image
This is very useful to discover other subsections, but it's not showing up in the new version.
old: https://astropy--14477.org.readthedocs.build/en/14477/user/units/quantity.html
new: https://astropy--14867.org.readthedocs.build/en/14867/units/quantity.html

**Comment by saimn** [comment]
That must be because of the new levels of toctrees. I think the navigation between subsections/subpackages with the sidebar was clearer in the old version. 
Also, 
- not sure if we need the sidebar in the index page, it duplicates the content of the page
- a lot of the space in the index page is taken by less important things (e.g. the Development section) whereas the most important (the list of subpackages) is hidden.

**Comment by pllim** [comment]
> Probably the only time I'll know enough to contribute to astropy

Don't be silly, @drammock ! 😸

**Comment by taldcroft** [comment]
My initial review of the output is that this looks fantastic. I strongly endorse moving forward with this effort!

**Comment by saimn** [comment]
> I struggle with this one because it is tough to retain the old behavior without moving the files around (as you did, but the it got conflicts very quickly). If we were to do it your way, this PR will have to get merged really fast after I make move the files. What do you think?

I think the most important is that the docs are well organized and easy to browse, for users. So if we need to move files for that and we don't have a good alternative, then we just need to do it. If there is an agreement then we can do that quickly enough to avoid conflicts I guess.

**Comment by pllim** [comment]
@saimn , I think I figured it out. It was a matter of where to put the headings and how many levels of sub-headings to expose in TOC. A little tricky but I think the layout now closely matches what you came up with in #14477 without moving files around, at least when I built it locally. Let's see if RTD agrees.

**Comment by pllim** [comment]
Ugh, there's one more thing... functions from automodapi is showing up in left menu

https://astropy--14477.org.readthedocs.build/en/14477/user/units/index.html

vs

https://astropy--14867.org.readthedocs.build/en/14867/units/index.html

**Comment by pllim** [comment]
pre-commit.ci autofix

**Comment by pllim** [comment]
@saimn , I think I fixed the TOC. Can you please have another look and let me know if I missed anything? Thanks!

**Comment by pllim** [comment]
This is as good as I can make it look with my limited Sphinx-fu. I released sphinx-astropy today, so this is now ready for review. Thanks!

**Comment by taldcroft** [comment]
@pllim - thanks again for working on this!

It appears that on the main docs landing page the User Guide navigation sections are not showing up. I'm referring to the `Data structures and transformations`, `Files, I/O, and Communication`, etc subsections that are pretty crucial for navigating the sub-packages. Interestingly they *are* there in the left sidebar once you get into the user guide.

One option is to go to a simpler landing page that does not have all the sub-sub-sections listed. I.e. make it look more like https://numpy.org/doc/stable/index.html or https://pandas.pydata.org/docs/. 

Also, the astropy favicon is not being show in the browser tab label.

**Comment by hamogu** [comment]
I like it very much overall and I'll be using this in the next couple of days to see more how it works in practice. 

The one thing that I find irritating is on the landing page. Because it has no section navigation, the main text it on the left of the page. Then, when I click any link to anything, it suddenly jumps to the middle, because the section navigation appears. It's not a deal breaker and I'm certainly not a trained UI expert, but personally I would prefer to either put some content in the section navigation (just so it's there all the time and does not jump in and out) or just have an empty space there (for example, numpy puts the search bar in there: https://numpy.org/doc/stable/index.html)

**Comment by pllim** [comment]
Re: Doc navigation -- I am trying to accomplish this without moving all the files around, so there is limitation of what it can do. Numpy/Pandas probably had a professional do it for them. I can see if there is something I can copy over, but if not, we might wanna consider chucking that as "future work" with Google Season of Docs or something.

**Comment by taldcroft** [comment]
@hamogu - I don't feel too strongly about the content of the astropy core docs landing page, though I disagree a little about https://astropy.org being the simplified version for the astropy core docs. The main Astropy Project page is about the Project, affiliated packages, team etc etc, as opposed to the documentation for the single astropy core package.

Part of what bothers me is that the pydata-sphinx-theme was not quite designed with the current structure in mind. There are two ways to view the User guide contents:
- https://astropy--14867.org.readthedocs.build/en/14867/#user-guide
- https://astropy--14867.org.readthedocs.build/en/14867/index_user_docs.html
That duality, and the fact that the subsections like `Files, I/O etc` are not in the first one, is what bothers my OCD a bit. There should just be one place to view the user guide contents.

**Comment by pllim** [comment]
Re: duality -- I *think* what you are asking for is to split each of the categories to a different RST sub-page. I tried that in a previous iteration but @saimn didn't like how it looked. I guess I can try again but I'd rather not go back and forth if someone else comes in and wants to revert.

### Let's make a decision right now before I push anymore commits though

1. What actually blocks this PR?
2. What is just nice-to-have that we can chuck as future work?
3. Or you rather just keep the old theme?

**Comment by hamogu** [comment]
I go with option 2 - can merge and tweak later if needed.

**Comment by pllim** [comment]
p.s. I should probably fix the tab icon, but logo and nice looking cards, I am not sure.

**Comment by hamogu** [comment]
[https://astropy--14867.org.readthedocs.build/en/14867/index_user_docs.html](https://astropy--14867.org.readthedocs.build/en/14867/index_user_docs.html) - that's the version that does have a side bar on the left and the main content stays in place when I click. Good enough for me (and I'm find with the other version also existing).

Now, who gets to review this and tick it off? Any maintainer (because it touches all sub-packages)? The documentation infrastructure maintainer (though this is not all about infrastructure)? A general maintainer?

Not clear to me, so I suggest we collect a few approving reviews and then merge.

**Comment by saimn** [comment]
> It appears that on the main docs landing page the User Guide navigation sections are not showing up. [...] 
One option is to go to a simpler landing page that does not have all the sub-sub-sections listed. I.e. make it look more like https://numpy.org/doc/stable/index.html or https://pandas.pydata.org/docs/.

Agreed, the sections are present in https://astropy--14867.org.readthedocs.build/en/14867/index_user_docs.html and in the sidebar, but cannot be exported to the top-level index page. So a solution is to have a simple landing page like numpy/scipy (we can copy it).

> Re: duality -- I think what you are asking for is to split each of the categories to a different RST sub-page. I tried that in a previous iteration but @saimn didn't like how it looked.

No I think @taldcroft mean to avoid the duplication between 
    https://astropy--14867.org.readthedocs.build/en/14867/#user-guide
    https://astropy--14867.org.readthedocs.build/en/14867/index_user_docs.html
by making the first one simpler (like https://numpy.org/doc/stable/)

> The one thing that I find irritating is on the landing page. Because it has no section navigation, the main text it on the left of the page. Then, when I click any link to anything, it suddenly jumps to the middle, because the section navigation appears.

Indeed, we could put the search bar on the left so the size will not change between main index and other pages. But maybe with the simpler index page it would not be such a problem ?

**Comment by hamogu** [comment]
> >  The one thing that I find irritating is on the landing page. Because it has no section navigation, the main text it on the left of the page. Then, when I click any link to anything, it suddenly jumps to the middle, because the section navigation appears.

> Indeed, we could put the search bar on the left so the size will not change between main index and other pages. But maybe with the simpler index page it would not be such a problem ?

I think so.

**Comment by taldcroft** [comment]
@saimn got it exactly right in interpreting my comments. Personally I prefer the simpler (numpy-style) landing page over the current situation. IMO seeing the following on the main astropy docs landing page is a blocker. Basically the first thing users see is something that is not useful. 

image

**Comment by pllim** [comment]
Logo is indeed much nicer. Thanks, @saimn !

File: docs/index.rst
Comment (pllim) [inline docs/index.rst:12]: @saimn , why you put version here? RTD gives you the version already, no?

**Comment by saimn** [comment]
I modified the index page to use cards (copied from Numpy and some bits from pandas).
https://astropy--14867.org.readthedocs.build/en/14867/
We don't have an "API Reference" section (since those are directly in the subpackages docs) so I replaced that card with the "Project details" section.

File: docs/index.rst
Comment (saimn) [reply docs/index.rst:12]: we have it currently on the top left, and I think it's useful to have it visible somewhere to avoid confusion, so I copied Numpy's way to show it.

**Comment by pllim** [comment]
@taldcroft / @hamogu , does it look better now?

Still need to figure out the favicon though.

**Comment by hamogu** [comment]
I like it.

**Comment by taldcroft** [comment]
I like it very much!!

**Comment by pllim** [comment]
pre-commit.ci autofix

**Comment by pllim** [comment]
OMG pre-commit went overboard... why did it fix the icons?! I have to drop that commit.

File: .ruff.toml
Comment (pllim) [inline .ruff.toml:321]: @nstarman or @eerovaher , this didn't work. How do I ask pre-commit.ci to ignore files in `docs/_static`?

File: .pre-commit-config.yaml
Comment (pllim) [inline .pre-commit-config.yaml:40]: This didn't work either to make pre-commit.ci pass 😿

**Comment by pllim** [comment]
Okay, so I think these feedback are implemented now (thank you very much to @saimn!):

✔️ Fancy cards on index page.
✔️ Logo looks better in dark mode.
✔️ Tab icon.
❓ Something about moving the search bar, but the a follow-up comment said we don't need this if we put in fancy cards on index page. Can you please confirm that this request is no longer relevant?

**Comment by hamogu** [comment]
Confirm that moving search bar is no longer needed, because we now have fancy index cards.

File: .ruff.toml
Comment (nstarman) [reply .ruff.toml:321]: This won't work because we'd have to add every rule to this list. The idea above to add to pre-commit's exclude list is the right approach.

Comment (pllim) [reply .ruff.toml:321]: @nstarman , I don't speak the language of ruff or pre-commit, can you pls tell me exactly what I should do here? Thanks!

**Comment by mhvk** [comment]
Overall, there are many things I like about the new theme, especially the separate left and right hand bars, which expand nicely. Main grumpiness is the waste of vertical space (see below), but I suppose that can be fixed after the fact by tweaking settings. (Obviously, only if people agree that it is bad; I still find the same with black: even though I'm getting used to it formatting, it is unfortunate one looses overview of the code).

Comparison old & new -- latter already shrunk to 90% to keep the font size of regular text similar. Note how much space is wasted by non-informative shadows around boxes, etc.; in the old version, you at least get to see the start of actual information and a first example.

image
image

**Comment by pllim** [comment]
@mhvk , I think that is customizable but I also find it tricky to find the correct font size and spacing that everyone can agree on. I actually find the new style easier on the eyes myself, and I don't mind the extra space because you can always jump to "Usage" from the sidebar on the right.

Regardless, feel free to tweak values mentioned in https://pydata-sphinx-theme.readthedocs.io/en/stable/user_guide/styling.html#base-variables and see if you find something that works better. 🤞

**Comment by pllim** [comment]
@mwcraig what do you mean by the "old documentation"?

<hr/>

That looks beautiful -- took me a moment or two to figure out how to get back to the "old" documentation, so when the change goes live we may want to send out an announcement or add to the appropriate card that it is the "old" documentation.
 
Thanks,
Matt
 
PS 😍 dark mode

**Comment by pllim** [comment]
As discussed at 2023-07-07 telecon, there was no objection to push this forward among the attendees, so I have rebased this PR and now going to put this on CoCo's agenda for final approval.

**Comment by mwcraig** [comment]
Looking forward to this quite a bit -- thanks to everyone who made it happen!

**Review by WilliamJamieson — APPROVED** [review]

**Comment by eteq** [comment]
In general I'm liking this now, but one question: is the intent to keep all the old docs as they were, and switch to this on dev (and then presumably it's "offical" in the next release)? Or to try to push the theme back to the older versions as well?

I'm in favor of it "from now on" rather than trying to change any older versions, which I *think* is the intent, but just confirming.

**Comment by eteq** [comment]
Ok awesome, then I like where it is now!

**Review by nstarman — APPROVED** [review]
I agree. Great to get this out and iterate from there. Thanks @pllim!

**Comment by pllim** [comment]
It failed with a warning that is unrelated but it does block RTD from building. 😭 

`WARNING: The following HTTPError has occurred fetching https://numpy.org/doc/stable//: 404 (Not Found)`

**Comment by pllim** [comment]
RTD passed but didn't report back. I am just gonna merge manually. Thanks!

**Comment by dhomeier** [comment]
Thanks for getting this done! I do agree to some extent with https://github.com/astropy/astropy/pull/14867#issuecomment-1582640075, although having both Notes and lengthy Warnings is a particularly disadvantageous example, and I also find the box rendering in the new docs much more pleasant on the eyes.
Section headings font sizes indeed seem quite excessive, so there probably is some fine-tuning to be done in the follow-up.

**Comment by pllim** [comment]
Most section heading levels are inherited from the old theme, so I think modifying the CSS rather than changing the RST would probably be less disruptive. I will open issue.

**Pull request merged**
